common/tier4_autoware_utils/include/tier4_autoware_utils/ros/published_time_publisher.hpp

```diff
@@ -0,0 +1,131 @@
+  }
+};
+
+struct GidEqual
+{
+  bool operator()(const rmw_gid_t & lhs, const rmw_gid_t & rhs) const noexcept
+  {
+    return std::memcmp(lhs.data, rhs.data, RMW_GID_STORAGE_SIZE) == 0;
+  }
+};
+
+class PublishedTimePublisher
+{
+public:
+  static std::unique_ptr<PublishedTimePublisher> create(
+    rclcpp::Node * node, const std::string & end_name = "/debug/published_time",
+    const rclcpp::QoS & qos = rclcpp::QoS(1))
+  {
+    const bool use_published_time = node->declare_parameter<bool>("use_published_time", false);
+    if (use_published_time) {
+      return std::unique_ptr<PublishedTimePublisher>(
+        new PublishedTimePublisher(node, end_name, qos));
+    } else {
+      return nullptr;
+    }
+  }
+
+  void publish(const rclcpp::PublisherBase::ConstSharedPtr & publisher, const rclcpp::Time & stamp)
+  {
+    const auto & gid = publisher->get_gid();
+    const auto & topic_name = publisher->get_topic_name();
+
+    // if the publisher is not in the map, create a new publisher for published time
+    if (publishers_.find(gid) == publishers_.end()) {
+      publishers_[gid] = node_->create_publisher<PublishedTime>(
+        static_cast<std::string>(topic_name) + end_name_, qos_);
+    }
+
+    const auto & pub_published_time_ = publishers_[gid];
+
+    // Check if there are any subscribers, otherwise don't do anything
+    if (pub_published_time_->get_subscription_count() > 0) {
+      PublishedTime published_time;
+
+      published_time.header.stamp = stamp;
+      published_time.published_stamp = rclcpp::Clock().now();
+
+      pub_published_time_->publish(published_time);
+    }
+  }
+
+  void publish(
+    const rclcpp::PublisherBase::ConstSharedPtr & publisher, const std_msgs::msg::Header & header)
+  {
+    const auto & gid = publisher->get_gid();
+    const auto & topic_name = publisher->get_topic_name();
+
+    // if the publisher is not in the map, create a new publisher for published time
+    if (publishers_.find(gid) == publishers_.end()) {
+      publishers_[gid] = node_->create_publisher<PublishedTime>(
+        static_cast<std::string>(topic_name) + end_name_, qos_);
+    }
+
+    const auto & pub_published_time_ = publishers_[gid];
+
+    // Check if there are any subscribers, otherwise don't do anything
+    if (pub_published_time_->get_subscription_count() > 0) {
+      PublishedTime published_time;
+
+      published_time.header = header;
+      published_time.published_stamp = rclcpp::Clock().now();
+
+      pub_published_time_->publish(published_time);
+    }
+  }
+
+private:
+  explicit PublishedTimePublisher(
+    rclcpp::Node * node, const std::string & end_name, const rclcpp::QoS & qos)
+  : node_(node), end_name_(end_name), qos_(qos)
+  {
+  }
+
+  rclcpp::Node * node_;
+  std::string end_name_;
+  rclcpp::QoS qos_;
+
+  // store them for each different publisher of the node
+  std::unordered_map<rmw_gid_t, rclcpp::Publisher<PublishedTime>::SharedPtr, GidHash, GidEqual>
```

Comment:
I've analyzed the overall implementation and it is really well thought out.

However, this overloaded hash and equality operators make the code hard to read.

Suggestion:

use a simple unordered map like:
```cpp
std::unordered_map<std::string, rclcpp::Publisher<PublishedTime>::SharedPtr> publishers_;
```

have this function:
```cpp
std::string gid_to_str_key(const rmw_gid_t & gid) const {
  return std::string(reinterpret_cast<const char *>(gid.data), RMW_GID_STORAGE_SIZE);
}
```

and whoever wants to access the map uses this step:
```
const std::string gid_key = gid_to_str_key(publisher->get_gid());
publishers_.find(gid_key) ...
publishers_[gid_key] = ...
```

This makes it overall more readable.